src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder_ListPatterns.cs

```diff
@@ -10,40 +10,38 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     internal sealed partial class DecisionDagBuilder
     {
-        private void MakeTestsAndBindingsForLengthAndListPatterns(BoundDagTemp input, BoundRecursivePattern recursive, ArrayBuilder<BoundPatternBinding> bindings, ArrayBuilder<Tests> tests)
+        private Tests MakeTestsAndBindingsForListPattern(BoundDagTemp input, BoundListPattern list, out BoundDagTemp output, ArrayBuilder<BoundPatternBinding> bindings)
         {
-            if (recursive.HasErrors)
-                return;
+            Debug.Assert(input.Type.IsErrorType() || list.HasErrors || list.InputType.IsErrorType() ||
+                         input.Type.Equals(list.InputType, TypeCompareKind.AllIgnoreOptions) &&
+                         list.Subpatterns.Count(p => p.Kind == BoundKind.SlicePattern) <= 1 &&
+                         list.LengthProperty is not null);
 
-            Debug.Assert(recursive.LengthProperty is not null || input.Type.IsSZArray());
-            var syntax = recursive.Syntax;
-            var lengthProperty = recursive.LengthProperty ?? (PropertySymbol)_compilation.GetSpecialTypeMember(SpecialMember.System_Array__Length);
-            var lengthEvaluation = new BoundDagPropertyEvaluation(syntax, lengthProperty, input);
-            tests.Add(new Tests.One(lengthEvaluation));
-            var lengthTemp = new BoundDagTemp(syntax, _compilation.GetSpecialType(SpecialType.System_Int32), lengthEvaluation);
+            var syntax = list.Syntax;
+            var subpatterns = list.Subpatterns;
+            var tests = ArrayBuilder<Tests>.GetInstance(4 + subpatterns.Length * 2);
+            output = input = MakeConvertToType(input, list.Syntax, input.Type.StrippedType(), isExplicitTest: false, tests);
```

Comment:
nit: consider also assertion that the stripped type of `input` is the narrowed type of `list`.